### PR TITLE
[FLINK-25404] Let blob.storage.directory default to <WORKING_DIR>/blobStorage

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

This PR lets the `BlobServer` and `BlobCache` use the `<WORKING_DIR>/blobStorage` directory by default. In order to facilitate the proper clean up of the blob storage directory, a `Reference` utility was introduced that allows to express whether the directory should be cleaned up by the `BlobServer`/`BlobCache` or the owning class.

This PR is based on #18083.

## Verifying this change

This change is mostly covered by existing tests for the BlobServer/BlobCache. Only tests for the newly introduced `Reference` utility were added.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobManagerSharedServicesTest.java
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobManagerSharedServicesTest.java:115]: shouldn't this use a proper tmp directory?

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java:185]:
```suggestion
                    "Could not create storage directory for BLOB store because no storage directory has been specified.");
```

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java:185]: maybe also include the corresponding config key

**Review by zentol — APPROVED** [review]

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java:185]: Will do.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobManagerSharedServicesTest.java
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobManagerSharedServicesTest.java:115]: Yes, I think so. Will update it.

**Comment by tillrohrmann** [comment]
Thanks for the review @zentol. I've addressed your comments. Will merge the PR once #18083 has been merged.

**Pull request closed without merging**
